distribution/lib/Standard/Base/0.0.0-dev/src/Data/Range.enso

```diff
@@ -313,8 +329,143 @@ type Range
+    index_of self element start=0 =
+        check_start_valid start self used_start->
+            case element of
+                _ : Integer -> get_index self used_start self.length-1 element
+                _ : Function ->
+                    result = find_internal self used_start element
+                    if result.is_nothing then Nothing else result.second
+                _ -> Error.throw (Illegal_Argument.Error "`Range.index_of` only accepts Integers or a predicate.")
+
+    ## Returns the last index of an element in the range.
+       Returns Nothing if the element is not found.
+
+       Arguments:
+       - element: The integer to search for or a predicate function to test for
+         each element.
+       - start: The index to start searching backwards from. If the index is
+         negative, it is counted from the end of the range.
+
+       > Example
+         Find the last index of an element in a pair.
+
+             Pair.new 2 2 . last_index_of 2 == 1
+    last_index_of : (Any | (Any -> Boolean)) -> Integer -> Integer | Nothing
+    last_index_of self element start=-1 =
+        if self.is_empty && (start==-1 || start==0) then Nothing else
+            check_start_valid start self used_start->
+                case element of
+                    _ : Integer -> get_index self 0 used_start element
+                    _ : Function ->
+                        start_value = self.start + used_start*self.step
+                        go current idx =
+                            if element current then idx else
+                                if current == self.start then Nothing else
+                                    @Tail_Call go current-self.step idx-1
+                        go start_value used_start
+                    _ -> Error.throw (Illegal_Argument.Error "`Range.last_index_of` only accepts Integers or a predicate.")
+
+    ## Reverses the range, returning a range with the same elements, but in
+       the opposite order.
+
+       > Example
+         Reverse a step 5 from 0 to 23
+
+             Range.new 0 23 . with_step 5 . reverse
+    reverse : Range
+    reverse self =
+        if self.step == 0 then throw_zero_step_error else
+            Range.Between self.last (self.first-self.step) (-self.step)
+
+    ## Converts the range to a vector containing the numbers in the range.
+
+       > Example
+         Getting a vector of the numbers 1 to 5.
+
+             1.up_to 6 . to_vector
+    to_vector : Vector
+    to_vector self = self.map x->x
+
+    ## Combines all the elements of a non-empty range using a binary operation.
+       If the range is empty, it throws an `Empty_Error`.
+
+       Arguments:
+       - function: A binary operation that takes two items and combines them.
+
+       > Example
+         Compute the sum of all the elements in a range.
+
+             0.up_to 10 . reduce (+)
+    reduce : (Any -> Any -> Any) -> Any ! Empty_Error
+    reduce self function =
```

Comment:
It might be useful to have `if_empty=Error.throw Empty_Error`. Then I could write `sum` as:
```
sum s = s.reduce (+) if_empty=0
```

Reply:
Good idea - will amend reduce.